tainting: pro: Basic index sensitivity

IagoAbal opened the pull request:
Make Semgrep track taint on individual indexes of a data structure, when the indexes are constant (integers or strings) and the code uses the built-in syntax for array indexing (typically `E[i]`).

```python
 x[1] = safe
 x[42] = source()
 sink(x) // no more finding
 sink(x[1]) // no more finding
 sink(x[42]) // finding
 sink(x[i]) // finding
 ```

Followed by PR semgrep/semgrep-proprietary#1318

test plan:
See semgrep/semgrep-proprietary#1318

IagoAbal committed:
tainting: pro: Basic index sensitivity

Followed by TODO

test plan:
See TODO

IagoAbal committed:
Fix test expectations for Pro

akuhlens commented:
I know this isn't how it currently works but is there anyway to make `sink(x)` match? I think this makes more sense from the standpoint of the aggregate giving you access to the tainted data. For instance most often sinking is logging, printing, or writing a value somewhere it doesn't belong. When you do this with an aggregate usually it writes the subcomponents.

akuhlens commented on src/il/IL_helpers.ml:
```diff
@@ -156,6 +157,19 @@ module LvalOrdered = struct
               | Dot a, Dot b -> compare_name a b
+              | Index { e = Operator ((G.Mult, _), []); _ }, Index _
+              | Index _, Index { e = Operator ((G.Mult, _), []); _ } ->
+                  (* This stands for an arbitrary index, like '[*]',
```
What context can this syntax originate from? The pattern language, python, somewhere else?

IagoAbal replied:
There is a `see 'Pro_taint_lval_env.lval_index_any'`, if you go see that definition and how it's used in that module, does it clarify it?

IagoAbal replied:
I expanded it a bit more, and I'll also add some more docs in `Pro_taint_lval_env`, let me now if that's not clear enough and I can improve it!

akuhlens commented on src/tainting/Dataflow_tainting.ml:
```diff
@@ -637,21 +638,41 @@ let find_pos_in_actual_args args_taints fparams =
+    | Some arg_offset_of_il_offset -> arg_offset_of_il_offset il_offset
+  in
+  (* TODO: Aren't we missing here C# and Go ? *)
+  if env.lang =*= Lang.Java || Lang.is_js env.lang || env.lang =*= Lang.Python
```
Would it make sense to check the list of pro-languages to keep this in sync as we add new languages to pro?

IagoAbal replied:
Nice idea, in principle we could even just do it for every language, but need to check if anything breaks.

IagoAbal replied:
I know you mean Pro-supported languages, because by "Pro language" we typically mean languages _only_ available in Pro like Apex and Elixir.

IagoAbal replied:
Let me put this in a separate PR just in case...

akuhlens commented on src/tainting/Dataflow_tainting.ml:
```diff
@@ -1269,7 +1288,34 @@ let lval_of_sig_arg fun_exp fparams args_exps (sig_arg : T.arg) :
   let os =
-    sig_arg.offset |> List_.map (fun x -> { o = Dot x; oorig = NoOrig })
+    sig_arg.offset
+    |> List_.map (fun x ->
+           match x with
```
You could use `function` here.

akuhlens commented on src/tainting/Taint.ml:
```diff
@@ -152,14 +155,18 @@ let _show_base base =
-      os |> List_.map (fun n -> fst n.IL.ident) |> String.concat "."
-    in
-    "." ^ os_str
+    let os_str = os |> List_.map _show_offset |> String.concat "" in
+    os_str
```
```suggestion
    os |> List_.map _show_offset |> String.concat ""
```

akuhlens reviewed (approved):
🙈

IagoAbal commented:
> I know this isn't how it currently works but is there anyway to make `sink(x)` match? I think this makes more sense from the standpoint of the aggregate giving you access to the tainted data. For instance most often sinking is logging, printing, or writing a value somewhere it doesn't belong. When you do this with an aggregate usually it writes the subcomponents.

Yeah this wasn't well considered when we added field sensitivity. The question was raised before, but because it's not raised very often I have not make it a priority. It wouldn't be hard to do, although I would have a dedicated PR for it. (This will also be easier if we add "taint shapes" which I'm likely to add right after this PR is merged.) I added a Linear ticket for this.

IagoAbal committed:
Cleanup

IagoAbal committed:
Andre's review
